### PR TITLE
fix(schemaorg): Enumeration type detection improved

### PR DESCRIPTION
**Pull request by stiivo**
Enumeration type detection improved by traversing the graph and checking if one of the parent elements is of type schema.org/Enumeration

| Q             | A
| ------------- | ---
| Branch?       | main
| Tickets       | #400
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Schema/Model/Class_.php [hunk 1]
```diff
@@ -66,10 +66,18 @@ public function getSubClassOf(): array
         return array_filter($this->resource->all('rdfs:subClassOf', 'resource'), static fn (RdfResource $resource) => !$resource->isBNode());
     }
 
-    public function isEnum(): bool
+    /**
+     * @param $resource
+     * @return bool
+     */
```
Comment (alanpoulain) [inline src/Schema/Model/Class_.php:72]: The PHPDoc is not needed.

File: src/Schema/Model/Class_.php [hunk 2]
```diff
@@ -66,10 +66,18 @@ public function getSubClassOf(): array
...
+    public function isEnum($resource = null): bool
     {
-        $subClassOf = $this->resource->get('rdfs:subClassOf');
+        if (is_null($resource)) {
+            $parentClass = $this->resource->get('rdfs:subClassOf');
+        } else {
+            $parentClass = $resource->get('rdfs:subClassOf');
+        }
```
Comment (alanpoulain) [inline src/Schema/Model/Class_.php:79]: To be shorter:
```suggestion
        $parentClass = ($resource ?? $this->resource)->get('rdfs:subClassOf');
```

File: src/Schema/Model/Class_.php [hunk 3]
```diff
@@ -66,10 +66,18 @@ public function getSubClassOf(): array
         return array_filter($this->resource->all('rdfs:subClassOf', 'resource'), static fn (RdfResource $resource) => !$resource->isBNode());
     }
 
-    public function isEnum(): bool
+    /**
+     * @param $resource
+     * @return bool
+     */
+    public function isEnum($resource = null): bool
```
Comment (alanpoulain) [inline src/Schema/Model/Class_.php:73]:
```suggestion
    public function isEnum(RdfResource $resource = null): bool
```

File: src/Schema/Model/Class_.php [hunk 4]
```diff
@@ -66,10 +66,18 @@ public function getSubClassOf(): array
...
     {
-        $subClassOf = $this->resource->get('rdfs:subClassOf');
+        if (is_null($resource)) {
+            $parentClass = $this->resource->get('rdfs:subClassOf');
+        } else {
+            $parentClass = $resource->get('rdfs:subClassOf');
+        }
 
-        return $subClassOf && self::SCHEMA_ORG_ENUMERATION === $subClassOf->getUri();
+        return !is_null($parentClass) && ($parentClass->getUri() === self::SCHEMA_ORG_ENUMERATION || $this->isEnum($parentClass));
```
Comment (alanpoulain) [inline src/Schema/Model/Class_.php:81]:
```suggestion
        return $parentClass && ($parentClass->getUri() === self::SCHEMA_ORG_ENUMERATION || $this->isEnum($parentClass));
```

**Comment by alanpoulain** [comment]
Thanks for the PR :slightly_smiling_face:

You would want to add some tests too (E2E and/or unit).

**Comment by dunglas** [comment]
Thanks @stiivo

**Pull request merged**